src/librustc_mir/build/scope.rs

```diff
@@ -822,6 +828,75 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         span_bug!(span, "region scope {:?} not in scope to drop {:?}", region_scope, local);
     }
 
+    /// Indicates that the "local operand" stored in `local` is
+    /// *moved* at some point during execution (see `local_scope` for
+    /// more information about what a "local operand" is -- in short,
+    /// it's an intermediate operand created as part of preparing some
+    /// MIR instruction). We use this information to suppress
+    /// redundant drops on the non-unwind paths. This results in less
+    /// MIR, but also avoids spurious borrow check errors
+    /// (c.f. #64391).
+    ///
+    /// Example: when compiling the call to `foo` here:
+    ///
+    /// ```rust
+    /// foo(bar(), ...)
+    /// ```
+    ///
+    /// we would evaluate `bar()` to an operand `_X`. We would also
+    /// schedule `_X` to be dropped when the expression scope for
+    /// `foo(bar())` is exited. This is relevant, for example, if the
+    /// later arguments should unwind (it would ensure that `_X` gets
+    /// dropped). However, if no unwind occurs, then `_X` will be
+    /// unconditionally consumed by the `call`:
+    ///
+    /// ```
+    /// bb {
+    ///   ...
+    ///   _R = CALL(foo, _X, ...)
+    /// }
+    /// ```
+    ///
+    /// However, `_X` is still registered to be dropped, and so if we
+    /// do nothing else, we would generate a `DROP(_X)` that occurs
+    /// after the call. This will later be optimized out by the
+    /// drop-elaboation code, but in the meantime it can lead to
+    /// spurious borrow-check errors -- the problem, ironically, is
+    /// not the `DROP(_X)` itself, but the (spurious) unwind pathways
+    /// that it creates. See #64391 for an example.
```

Comment:
Should we try to find other ways to solve this problem? I was of the impression that we aggressively avoided "cleverness" in the MIR building code.

Reply:
Also, this entire change seems like it could be its own entirely separate PR - am I missing something?

Reply:
Oh, seems to be https://github.com/rust-lang/rust/pull/64584#issuecomment-533192803.

Reply:
> Should we try to find other ways to solve this problem? I was of the impression that we aggressively avoided "cleverness" in the MIR building code.

Agreed -- still, I felt this was not overly clever. It seems kind of .. excessively dumb to create intermediates for use in the "call" (or whatever) that you are building that you then unconditionally move and -- immediately thereafter -- unconditionally free.

Overall though I think this "drop" code is a too complex, though I'm not sure how to improve it just now.